pkg/metadata/manager.go

```diff
@@ -0,0 +1,374 @@
+		// Example:containerd://e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0e
+		_, after, found := strings.Cut(cId, "//")
+		if !found || after == "" {
+			log.Errorf("unexpected container ID: %s", cId)
+			continue
+		}
+		cId = after
+		if _, ok := pm.containers[cId]; ok {
+			currContainers[cId] = struct{}{}
+			// Container cgroup path doesn't change as long as the container id
+			// is the same.
+			continue
+		}
+		pm.containers[cId] = struct{}{}
+		currContainers[cId] = struct{}{}
+
+		// Container could've been gone, so don't log any errors.
+		cgrpPath, err := m.getContainerCgroupPath(id, cId, pod.Status.QOSClass)
+		if err != nil {
+			log.Debugf("failed to get container metadata for (%s): %v", cId, err)
+			continue
+		}
+		cgrpId, err := m.cgroupsChecker.GetCgroupID(cgrpPath)
+		if err != nil {
+			log.Debugf("failed to get cgroup id for cgroup path (%s): %v", cgrpPath, err)
+			continue
+		}
+		log.Infof("aditi-debug-metadata %s %d", cgrpPath, cgrpId)
+		m.containerMetadataByCgrpId[cgrpId] = &containerMetadata{
+			cgroupId:    cgrpId,
+			cgroupPath:  cgrpPath,
+			parentPodId: id,
+		}
+	}
+	// Clean up any pod's old containers.
+	if oldPod != nil {
+		for _, c := range oldPod.Status.ContainerStatuses {
+			if _, ok := currContainers[c.ContainerID]; !ok {
+				delete(pm.containers, c.ContainerID)
+			}
+		}
+	}
+}
+
+func (m *CgroupManager) deletePodMetadata(pod *v1.Pod) {
+	id := string(pod.ObjectMeta.UID)
+
+	if _, ok := m.podMetadataById[id]; !ok {
+		return
+	}
+	for k, cm := range m.containerMetadataByCgrpId {
+		if cm.parentPodId == id {
+			delete(m.containerMetadataByCgrpId, k)
+		}
+	}
+	delete(m.podMetadataById, id)
+}
+
+func (m *CgroupManager) getParentPodMetadata(cgroupId uint64, podMetadataOut chan *PodMetadata) {
+	cm, ok := m.containerMetadataByCgrpId[cgroupId]
+	if !ok {
+		log.Debugf("Metadata not found for container: %d", cgroupId)
+		close(podMetadataOut)
+		return
+	}
+
+	pm, ok := m.podMetadataById[cm.parentPodId]
+	if !ok {
+		log.Debugf("Parent pod metadata not found for container: %d", cgroupId)
+		close(podMetadataOut)
+		return
+	}
+	podMetadata := PodMetadata{
+		Name:      pm.name,
+		Namespace: pm.namespace,
+	}
+	podMetadata.IPs = append(podMetadata.IPs, pm.ips...)
+	log.Debugf("Parent pod metadata for container (%d): %+v", cgroupId, podMetadata)
+
+	podMetadataOut <- &podMetadata
+	close(podMetadataOut)
+}
+
+func (m *CgroupManager) baseCgroupPathForQos(path string, qos v1.PodQOSClass) string {
+	if qos == v1.PodQOSGuaranteed {
+		return fmt.Sprintf("%s", path)
+	}
+	return fmt.Sprintf("%s/%s", path, strings.ToLower(string(qos)))
+}
+
+func (m *CgroupManager) getContainerCgroupPath(podId string, containerId string, containerQos v1.PodQOSClass) (string, error) {
+	if m.templateCgroupBasePath == "" {
+		return "", fmt.Errorf("failed to get cgroup path for (%s)", containerId)
+	}
+
+	return fmt.Sprintf("%s/pod%s/%s", m.baseCgroupPathForQos(m.templateCgroupBasePath, containerQos),
```

Comment:
Is this guaranteed to always be the case? Even for cri-o?

Reply:
We are using known defaults that are also hard-coded in Kubernetes source code. What about cri-o -  is the base cgroup path different?

Reply:
I'm not sure, but the path does look different. Looking a a container in a pod, I see this path (on the host):

`/sys/fs/cgroup/systemd/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podaada424e_b978_437b_bd12_5e22626a3c08.slice/crio-9b6be68b035f1b3453b0a0deca929d6c1953ce9bc9b4f18e6a44a9d8868e0328.scope`

Reply:
Can you post output of `sudo crictl inspect <container-id> | grep cgroupsPath`? If you have a code reference for Kubelet, then it'll be easy to adapt the logic in case crio has a different cgroup path.